tools/kvmexit.py

```diff
@@ -0,0 +1,393 @@
+            if (initial == NULL) {
+                return -1;
+            }
+
+            pcpu_kvm_stat.update(&cur_pid_tgid, initial);
+            tmp_info = pcpu_kvm_stat.lookup(&cur_pid_tgid);
+            // To pass the verifier
+            if (tmp_info == NULL) {
+                return -1;
+            }
+        }
+    }
+
+    if (er < REASON_NUM) {
+        tmp_info->exit_ct[er]++;
+        if (cache_miss == 1) {
+            if (cache_p->cache_pid_tgid != 0) {
+                // b.*.a Let's save the last hit cache_info into kvm_stat.
+                pcpu_kvm_stat.update(&cache_p->cache_pid_tgid, &cache_p->cache_exit_ct);
+            }
+            // b.* As the cur_pid_tgid meets current pcpu_cache_array for the first time, save it.
+            cache_p->cache_pid_tgid = cur_pid_tgid;
+            bpf_probe_read(&cache_p->cache_exit_ct, sizeof(*tmp_info), tmp_info);
+        }
+        return 0;
+    }
+
+    return -1;
+}
+"""
+
+# format output
+exit_reasons = (
+    "EXCEPTION_NMI",
+    "EXTERNAL_INTERRUPT",
+    "TRIPLE_FAULT",
+    "INIT_SIGNAL",
+    "N/A",
+    "N/A",
+    "N/A",
+    "INTERRUPT_WINDOW",
+    "NMI_WINDOW",
+    "TASK_SWITCH",
+    "CPUID",
+    "N/A",
+    "HLT",
+    "INVD",
+    "INVLPG",
+    "RDPMC",
+    "RDTSC",
+    "N/A",
+    "VMCALL",
+    "VMCLEAR",
+    "VMLAUNCH",
+    "VMPTRLD",
+    "VMPTRST",
+    "VMREAD",
+    "VMRESUME",
+    "VMWRITE",
+    "VMOFF",
+    "VMON",
+    "CR_ACCESS",
+    "DR_ACCESS",
+    "IO_INSTRUCTION",
+    "MSR_READ",
+    "MSR_WRITE",
+    "INVALID_STATE",
+    "MSR_LOAD_FAIL",
+    "N/A",
+    "MWAIT_INSTRUCTION",
+    "MONITOR_TRAP_FLAG",
+    "N/A",
+    "MONITOR_INSTRUCTION",
+    "PAUSE_INSTRUCTION",
+    "MCE_DURING_VMENTRY",
+    "N/A",
+    "TPR_BELOW_THRESHOLD",
+    "APIC_ACCESS",
+    "EOI_INDUCED",
+    "GDTR_IDTR",
+    "LDTR_TR",
+    "EPT_VIOLATION",
+    "EPT_MISCONFIG",
+    "INVEPT",
+    "RDTSCP",
+    "PREEMPTION_TIMER",
+    "INVVPID",
+    "WBINVD",
+    "XSETBV",
+    "APIC_WRITE",
+    "RDRAND",
+    "INVPCID",
+    "VMFUNC",
+    "ENCLS",
+    "RDSEED",
+    "PML_FULL",
+    "XSAVES",
+    "XRSTORS",
+    "N/A",
+    "N/A",
+    "UMWAIT",
+    "TPAUSE"
+)
+
+#
+# Do some check
+#
+
+try:
+    # Currently, only adapte on intel architecture
+    cmd = "cat /proc/cpuinfo | grep vendor_id | head -n 1"
+    arch_info = subprocess.check_output(cmd, shell=True).strip()
+    if b"Intel" in arch_info:
+        pass
+    else:
+        raise Exception("Currently we only support Intel architecture, please do expansion if needs more.")
+
+    # Check if kvm module is loaded
+    if os.access("/dev/kvm", os.R_OK | os.W_OK):
+        pass
+    else:
+        raise Exception("Please insmod kvm module to use kvmexit tool.")
+except Exception as e:
+    raise Exception("Failed to do precondition check, due to: %s." % e)
+
+try:
+    # Let's firstly try raw_tracepoint_module
+    if support_raw_tracepoint_module():
```

Comment:
This won't work. The correct way is `BPF.support_raw_tracepoint_in_module()`.

Reply:
> This won't work. The correct way is `BPF.support_raw_tracepoint_in_module()`.

Thanks for the reminder